packages/aio-commerce-lib-auth/source/lib/utils.ts

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright 2025 Adobe. All rights reserved.
+ * This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+import { allNonEmpty } from "@adobe/aio-commerce-lib-core/params";
+
+import { resolveImsAuthParams } from "./ims-auth/utils";
+import { resolveIntegrationAuthParams } from "./integration-auth/utils";
+
+import type { ImsAuthParams } from "./ims-auth/schema";
+import type { IntegrationAuthParams } from "./integration-auth/schema";
+
+/** IMS authentication parameters with the "ims" strategy indicator. */
+export type ImsAuthParamsWithStrategy = ImsAuthParams & { strategy: "ims" };
+
+/** Adobe Commerce Integration authentication parameters with the "integration" strategy indicator. */
+export type IntegrationAuthParamsWithStrategy = IntegrationAuthParams & {
+  strategy: "integration";
+};
+
+/**
+ * Union type representing either IMS or Integration authentication providers.
+ * Each provider includes its respective parameters and a strategy identifier.
+ */
+export type AuthProvider =
+  | ImsAuthParamsWithStrategy
+  | IntegrationAuthParamsWithStrategy;
+
+/**
+ * The available authentication resolver strategies.
+ * - `"ims"`: Force resolution as IMS authentication
+ * - `"integration"`: Force resolution as Adobe Commerce Integration authentication
+ * - `"auto"`: Automatically detect the authentication type based on provided parameters
+ */
+export type AuthResolverStrategy = AuthProvider["strategy"] | "auto";
+
+/** Conditional type that resolves to the correct authentication parameters type based on the strategy. */
+type ResolvedAuthParams<T extends AuthResolverStrategy> = T extends "ims"
+  ? ImsAuthParamsWithStrategy
+  : T extends "integration"
+    ? IntegrationAuthParamsWithStrategy
+    : AuthProvider;
+
+/**
+ * Automatically detects and resolves authentication parameters from App Builder action inputs.
+ * Attempts to resolve IMS authentication first, then falls back to Integration authentication.
+ *
+ * @param params The App Builder action inputs containing authentication parameters.
+ * @throws {Error} If neither IMS nor Integration authentication parameters can be resolved.
+ */
+function resolveAuthParamsAuto(params: Record<string, unknown>): AuthProvider {
+  if (
+    allNonEmpty(params, [
```

Comment:
Would make sense to me extracting the envvars in a const so we can use the array when building the error message